### PR TITLE
Sort expressions for computeAt when traversing expressions

### LINKED ISSUE #112: Invalid code generation
## 🐛 Bug

The testGPU_FusionSoftmax test in this branch generates invalid code:

https://github.com/naoyam/pytorch/tree/codegen-bug

Here's the generated kernel:

```
__global__ void CUDAGeneratedKernel(Tensor<float, 3> T0, Tensor<float, 3> T9){
  __shared__ float shared_mem[1024];
  float T6[1];
  float T5[1];
  T5[ 0 ]
     = float(0);
  float T11[1];
  if ( ( ( ( 0 * 32 ) + threadIdx.x ) < T9.size[2] ) ) {
    T11[ 0 ]
       = float(0);
  }
  float T2[1];
  float T1[1];
  T1[ 0 ]
     = float(0);
  float T10[1];
  if ( ( ( ( 0 * 32 ) + threadIdx.x ) < T9.size[2] ) ) {
    T10[ 0 ]
       = float(0);
  }
  for(size_t i84 = 0; i84 < ( ceilDiv(T9.size[2], 32) ); ++i84 ) {
    if ( ( ( ( i84 * 32 ) + threadIdx.x ) < T9.size[2] ) ) {
      T10[ 0 ]
         = fmaxf(T10[ 0 ]
        , T0[ ( blockIdx.x * T0.stride[0] ) + ( blockIdx.y * T0.stride[1] ) + ( ( ( i84 * 32 ) + threadIdx.x ) * T0.stride[2] ) ]);
    }
  }
  blockReduce< true, false, false > ( T1[ 0 ], T10[ 0 ], reduction_fmaxf_float, threadIdx, blockDim, reinterpret_cast<float*>(shared_mem));
  if ( ( threadIdx.x == 0 ) ) {
    T2[ 0 ]
       = T1[ 0 ];
  }
  for(size_t i87 = 0; i87 < ( ceilDiv(T9.size[2], 32) ); ++i87 ) {
    float T7[1];
    if ( ( ( ( i87 * 32 ) + threadIdx.x ) < T9.size[2] ) ) {
      T7[ 0 ]
         = T0[ ( blockIdx.x * T0.stride[0] ) + ( blockIdx.y * T0.stride[1] ) + ( ( ( i87 * 32 ) + threadIdx.x ) * T0.stride[2] ) ]
         - T2[ 0 ];
    }
    float T8[1];
    if ( ( ( ( i87 * 32 ) + threadIdx.x ) < T9.size[2] ) ) {
      T8[ 0 ]
         = expf(T7[ 0 ]);
    }
    float T3[1];
    if ( ( ( ( i87 * 32 ) + threadIdx.x ) < T9.size[2] ) ) {
      T3[ 0 ]
         = T0[ ( blockIdx.x * T0.stride[0] ) + ( blockIdx.y * T0.stride[1] ) + ( ( ( i87 * 32 ) + threadIdx.x ) * T0.stride[2] ) ]
         - T2[ 0 ];
    }
    float T4[1];
    if ( ( ( ( i87 * 32 ) + threadIdx.x ) < T9.size[2] ) ) {
      T4[ 0 ]
         = expf(T3[ 0 ]);
    }
  }
  for(size_t i93 = 0; i93 < ( ceilDiv(T9.size[2], 32) ); ++i93 ) {
    if ( ( ( ( i93 * 32 ) + threadIdx.x ) < T9.size[2] ) ) {
      T11[ 0 ]
         = T11[ 0 ]
         + T4[ 0 ];
    }
  }
  blockReduce< true, false, false > ( T5[ 0 ], T11[ 0 ], reduction_add_float, threadIdx, blockDim, reinterpret_cast<float*>(shared_mem));
  if ( ( threadIdx.x == 0 ) ) {
    T6[ 0 ]
       = T5[ 0 ];
  }
  for(size_t i96 = 0; i96 < ( ceilDiv(T9.size[2], 32) ); ++i96 ) {
    if ( ( ( ( i96 * 32 ) + threadIdx.x ) < T9.size[2] ) ) {
      T9[ ( blockIdx.x * T9.stride[0] ) + ( blockIdx.y * T9.stride[1] ) + ( ( ( i96 * 32 ) + threadIdx.x ) * T9.stride[2] ) ]
         = T8[ 0 ]
         / T6[ 0 ];
    }
  }
}
```

`T8` is used in the final loop nest but is defined and computed in a different loop nest.

The final loop nest corresponds to this expression:

```
TensorView* output_tv6 = div(exp_tv4_2, bcast_sum_tv6);
```

Interestingly, swapping the two operands, i.e., `div(bcast_sum_tv6, exp_tv4_2)`, seems to be fine. See https://github.com/naoyam/pytorch/blob/d2913c9fb2c7e94c786ce3b73a6b5ab87b4eee3b/test/cpp/jit/test_gpu.cpp#L2798. The fuser generates code as expected, although they are not the same computation anymore. 

Note that this issue was encountered in the block broadcast PR (#100) but shows up with the dev branch as well.

### PR DESCRIPTION
**Pull request by naoyam**
Loop-nest generation assumes that expressions are sorted in a topological order. However, it is not sufficient for computeAt to work in certain conditions. See #112 for the current limitation.

This PR fixes the problem by adjusting the topological order so that computeAt'ed expressions are visited last. This allows the expression of a tensor to be properly embedded in the loop nest of the target tensor even when there is also another no-computeAt input tensor (e.g., `T2` in https://github.com/csarofeen/pytorch/issues/112#issuecomment-648488887).

Sorting is only done when explicitly requested, although it is always valid to do. Because of this, an additional parameter, `respect_compute_at`, is added in several scattered places. Since it's always valid to do, the code can be simplified by simply enabling it always. Let me know if that'd be preferred.

This PR also simplifies the `traverseFrom` implementation, which I think would be easier to maintain. Let me know if the existing one should be kept. I'll revert the simplification if preferred.

Currently, there are a few environment variables just for debugging. I'll remove them before merging.

Fixes #112

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: nit: since inputs is an array of const Statement*, we can simply say inputs[i]->as<TensorView>(). It will still return a const TensorView* (the const is not needed)

File: torch/csrc/jit/codegen/cuda/fusion.h
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/fusion.h:157]: we could remove breadth_first - there's no implementation or use for it. If we ever want it it's easy to add it back.

@csarofeen , what do you think?

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: are you planning to remove the getenv before merging, or is this a TODO for sometime in the future?

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: Nit: instead of i0 and i1 can we use 'a' and 'b'? (easier to read, and somewhat common notation)

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: Simplified in what way? Does it preserve the semantics of traverseFrom()?

The name traverseFrom2() doesn't say much either, is it just temporary?

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: Yes, it's a TODO that I will do when making this PR non-draft.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: nit: the const is not really needed here

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: I'd recommend using { ... } even for single statements, any thoughts?

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: It should result in the same order. It is just temporary. I'm going to just keep the original one or the new one.

I find the original implementation has several places to improve mainly because part of the logic is replicated, e.g.,`handle` is called at multiple places. `next` is also called from two locations. That's what I wanted to clean up.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: I didn't want to change the original signature.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: I can't agree more, but this is also just copied from `traverseFrom`.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: Thanks. Didn't know that. Updated.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: I actually prefer `i0` and `i1` in this case since in the code below they are put into an array at the 0-th and 1st location, respectively.

File: test/cpp/jit/test_gpu.cpp
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp]: can you also check with just `tv2->computeAt(tv3...`) in this test?

File: torch/csrc/jit/codegen/cuda/fusion.h
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/fusion.h:157]: sure, sounds good.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp:39]: Could you add an error for `outputs().size() > 1`? Otherwise I think we should check all TVs if there's multiple outputs.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: nit: `stmt0` and `stmt1` just a small convention I've been using.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp:31]: nice modifications on next

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: is std::sort stable? I think this would give deterministic orderings, but just making sure.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: I'm lazy those keys are hard to hit :-P, just kidding, that's fine, I'll stop leaving them off.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: LGTM you can replace the original if you think this is cleaner.

**Review by csarofeen — CHANGES_REQUESTED** [review]
Overall changes to iter_visitor look good. I've got some API questions to consider:

Should expr order change from the users perspective when they call computeAt? This is more along the question of "do we set this traversal to default?".

If we decide no on the above, should it be an option from the user perspective?

If we don't think that's helpful, then could this just be a pass more separated from IterVisitor similar to the other lowering passes?

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: stm0 and stm1 are fine too. or first/second.

The problem with i1 is simply that i, l and 1 may look too similar with certain fonts.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: great. I agree that the new version looks cleaner.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: thanks! (even LLVM, which used to be hardcore on not using braces for single statements, is not reconsidering it)

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: good point. std::sort is not stable (but std::stable_sort() is)

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: So, do we want to use `std::stable_sort`? Both should be deterministic, so I think both are fine.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: Done.

File: test/cpp/jit/test_gpu.cpp
Comment (naoyam) [reply test/cpp/jit/test_gpu.cpp]: Added.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp:39]: Added.

File: torch/csrc/jit/codegen/cuda/fusion.h
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/fusion.h:157]: I'll take care of it later as a separate PR (#130)

**Comment by naoyam** [comment]
> Should expr order change from the users perspective when they call computeAt? This is more along the question of "do we set this traversal to default?".

The user should not care the traversal order of expression inputs. All that matters is to generate correct code. So, setting it as default should be fine.

> 
> If we decide no on the above, should it be an option from the user perspective?
> 
> If we don't think that's helpful, then could this just be a pass more separated from IterVisitor similar to the other lowering passes?

Making it separated from `IterVisitor` would be possible, but the seems to be a lot more difficult to implement than this implementation, where only `next()` is extended.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: I think it would only be deterministic if stably sorted no? Otherwise if multiple values have the same computeAt position, couldn't the order change between them?

**Comment by csarofeen** [comment]
```
Should expr order change from the users perspective when they call computeAt? This is more along the question of "do we set this traversal to default?".

  The user should not care the traversal order of expression inputs. All that matters is to generate correct code. So, setting it as default should be fine.
```
I actually don't think that's true, if I print the expressions of a fusion which I often do, it may be surprising to me that their order changes based on computeAt.

```
If we decide no on the above, should it be an option from the user perspective?

If we don't think that's helpful, then could this just be a pass more separated from IterVisitor similar to the other lowering passes?

Making it separated from IterVisitor would be possible, but the seems to be a lot more difficult to implement than this implementation, where only next() is extended.
```
I was thinking we could have a separate "GetSortedExprs" pass in the lowering process that would inherit from IterVisitor. This could be done as a follow up PR though depending on API questions which I'm not convinced of. As of now I think it's okay to move forward, but maybe we leave this ordering off by default for now.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: Yes, but that's still deterministic, isn't it?

**Comment by naoyam** [comment]
> ```
> Should expr order change from the users perspective when they call computeAt? This is more along the question of "do we set this traversal to default?".
> 
>   The user should not care the traversal order of expression inputs. All that matters is to generate correct code. So, setting it as default should be fine.
> ```
> 
> I actually don't think that's true, if I print the expressions of a fusion which I often do, it may be surprising to me that their order changes based on computeAt.

OK, that makes sense.

> 
> ```
> If we decide no on the above, should it be an option from the user perspective?
> 
> If we don't think that's helpful, then could this just be a pass more separated from IterVisitor similar to the other lowering passes?
> 
> Making it separated from IterVisitor would be possible, but the seems to be a lot more difficult to implement than this implementation, where only next() is extended.
> ```
> 
> I was thinking we could have a separate "GetSortedExprs" pass in the lowering process that would inherit from IterVisitor. This could be done as a follow up PR though depending on API questions which I'm not convinced of. As of now I think it's okay to move forward, but maybe we leave this ordering off by default for now.

OK, I agree. It's already off by default.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: While most implementation likely are, I don't think that std::sort() is guaranteed to be deterministic. If that's a concern here, using std::stable_sort() seems an easy solution.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: nit: please add { .. } here too

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: it seems we can skip this check and unconditionally push "from" to stmt_stack?

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: I'd use plain insert() here - we're inserting just a pointer, right? (emplace() suggests that there's an expensive construction going on)

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp:110]: no need to check if stmt is already visited?

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: I'm ok to just use `std::stable_sort`, but I'm curious what kind of non-deterministic algorithm is actually used.

Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: Unless it's documented (guaranteed) to be deterministic we shouldn't relay on that. Even if all current implementation happen to be deterministic, a future library change can break us.

PS. for a hypothetical source of non-determinism, let's say a std::sort() is implemented as a quick sort using a _random_ pivot choice.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp:110]: This should be fine. It's either taken care by `remove_visited` or multiple visits are indeed allowed.

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: OK, fair enough. I'll update it.

**Review by csarofeen — APPROVED** [review]

**Pull request merged**
